### PR TITLE
Fix connect bug

### PR DESCRIPTION
**Pull request by VirrageS**
Before the change, processes were not waiting for master even when they got
'connection refused' (master is not listening yet, so we should wait).
It was because we were closing socket twice: first, by
the resource guard; second, manually in exception handling.
That caused errno to be set to different value (9 - bad file descriptor)
and in result `if`, which checked if connection was refused, was failing.

### CODE REVIEW

File: torch/lib/THD/base/ChannelUtils.cpp [hunk 1]
```diff
@@ -181,28 +181,29 @@ int connect(const std::string& address, port_type port, bool wait, int timeout)
       }
 
       socklen_t err_len = sizeof(errno);
-      SYSCHECK(getsockopt(socket, SOL_SOCKET, SO_ERROR, &errno, &err_len));
+      ::getsockopt(socket, SOL_SOCKET, SO_ERROR, &errno, &err_len);
```
Comment (0mp) [inline torch/lib/THD/base/ChannelUtils.cpp]: Shouldn't you set errno to 0 beforehand?

File: torch/lib/THD/base/ChannelUtils.cpp [hunk 2]
```diff
@@ -181,28 +181,29 @@ int connect(const std::string& address, port_type port, bool wait, int timeout)
...
     } catch (std::exception& e) {
-      // if `connect` fails, the state of the socket is unspecified.
-      // we should close the socket and create a new one before attempting to reconnect.
-      ::close(socket);
       if (errno == ECONNREFUSED) any_refused = true;
 
-      // we need to move to next address because this was not available
-      // to connect or to create socket
+      // We need to move to next address because this was not available
+      // to connect or to create socket.
```
Comment (0mp) [inline torch/lib/THD/base/ChannelUtils.cpp]: a socket

File: torch/lib/THD/base/ChannelUtils.cpp [hunk 3]
```diff
@@ -181,28 +181,29 @@ int connect(const std::string& address, port_type port, bool wait, int timeout)
...
       break;
     } catch (std::exception& e) {
-      // if `connect` fails, the state of the socket is unspecified.
-      // we should close the socket and create a new one before attempting to reconnect.
-      ::close(socket);
       if (errno == ECONNREFUSED) any_refused = true;
 
-      // we need to move to next address because this was not available
-      // to connect or to create socket
+      // We need to move to next address because this was not available
```
Comment (0mp) [inline torch/lib/THD/base/ChannelUtils.cpp]: How about "the next address"?

File: torch/lib/THD/base/ChannelUtils.cpp [hunk 4]
```diff
@@ -181,28 +181,29 @@ int connect(const std::string& address, port_type port, bool wait, int timeout)
       }
 
       socklen_t err_len = sizeof(errno);
-      SYSCHECK(getsockopt(socket, SOL_SOCKET, SO_ERROR, &errno, &err_len));
+      ::getsockopt(socket, SOL_SOCKET, SO_ERROR, &errno, &err_len);
+      /* `errno` is set when:
+       *   1. `getsockopt` has failed
+       *   2. there was awaiting error in socket (we save error on `errno` variable)
```
Comment (0mp) [inline torch/lib/THD/base/ChannelUtils.cpp]: There is (s/an awaiting/a pending/ ?) error in the socket ...

File: torch/lib/THD/base/ChannelUtils.cpp (hunk 1 above)
Comment (0mp) [reply torch/lib/THD/base/ChannelUtils.cpp]: Oh I see, errno can be already set and we are about to throw an error.

File: torch/lib/THD/base/ChannelUtils.cpp (hunk 4 above)
Comment (0mp) [reply torch/lib/THD/base/ChannelUtils.cpp]: Something is not right here: "(we save error on `errno` variable)"

Maybe _There is (s/an awaiting/a pending/ ?) error in the socket we stored in the `errno` variable before._

File: torch/lib/THD/base/ChannelUtils.cpp (hunk 1 above)
Comment (apaszke) [reply torch/lib/THD/base/ChannelUtils.cpp]: No, errno is not set. We should check the return code or zero it

File: torch/lib/THD/base/ChannelUtils.cpp (hunk 4 above)
Comment (VirrageS) [reply torch/lib/THD/base/ChannelUtils.cpp]: What is actually not right here? Also I don't understand what are you trying to say in the proposal you gave.

File: torch/lib/THD/base/ChannelUtils.cpp (hunk 1 above)
Comment (VirrageS) [reply torch/lib/THD/base/ChannelUtils.cpp]: Hmm, the only place errno can be set is `::poll` but we check return code, so I guess this line always start with `errno = 0` or I miss something?

Comment (VirrageS) [reply torch/lib/THD/base/ChannelUtils.cpp]: Okay, that is not true. NVM!

File: torch/lib/THD/base/ChannelUtils.cpp (hunk 4 above)
Comment (0mp) [reply torch/lib/THD/base/ChannelUtils.cpp]: ! Sorry, I thought that those points (1., 2.) describe mutually exclusive events. As a result I thought that we've got a value of another error stored in `errno`. 

In that case my "we stored in the errno variable before." doesn't have any sense.

I'd change it to "we saved the error to the `errno` variable"

Comment (VirrageS) [reply torch/lib/THD/base/ChannelUtils.cpp]: Done

**Pull request merged**